src/gf.c

```diff
@@ -611,57 +609,73 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
-            if (elt == (jl_value_t*)jl_function_type)
-                jl_svecset(newparams, i, jl_any_type);
-        }
-        origtype = jl_apply_tuple_type(newparams);
-        temp = (jl_value_t*)origtype;
-    }
 
     // here we infer types and specialize the method
-    jl_array_t *lilist=NULL;
-    jl_lambda_info_t *li=NULL;
+    int from_specializations = 1;
     if (definition->specializations != NULL) {
         // reuse code already generated for this combination of lambda and
         // arguments types. this happens for inner generic functions where
         // a new closure is generated on each call to the enclosing function.
-        lilist = definition->specializations;
+        jl_array_t *lilist = definition->specializations;
         int k;
-        for(k=0; k < lilist->nrows; k++) {
-            li = (jl_lambda_info_t*)jl_array_ptr_ref(lilist, k);
-            if (jl_types_equal((jl_value_t*)li->specTypes, (jl_value_t*)type))
+        for (k = 0; k < lilist->nrows; k++) {
+            jl_lambda_info_t *li = (jl_lambda_info_t*)jl_array_ptr_ref(lilist, k);
+            if (jl_types_equal((jl_value_t*)li->specTypes, (jl_value_t*)type)) {
+                newmeth = li;
                 break;
+            }
         }
-        if (k == lilist->nrows) lilist=NULL;
     }
 
-    if (lilist != NULL && !li->inInference) {
-        assert(li);
-        newmeth = li;
-        jl_typemap_insert(cache, parent, type, jl_emptysvec, origtype, guardsigs, (jl_value_t*)newmeth, jl_cachearg_offset(mt), &lambda_cache, NULL);
+    if (!newmeth || newmeth->inInference) {
+        from_specializations = 0;
+        newmeth = jl_get_specialized(definition, type, sparams);
+    }
+
+    if (cache_with_orig) {
+        // if there is a need to cache with one of the original signatures,
+        // the method is still specialized on `types`,
+        // but one of the original types will be used as the entry signature
+        // in the method cache, possible with a simplesig also,
+        // to prevent anything else from matching this entry
+        type = origtype; // restore `type` to be the `origtype` backup (discard computed simplified `type`)
+        origtype = tt; // choose `tt` as the primary key
+        makesimplesig = 0;
+    }
+    else {
+        // don't need `origtype` anymore: `type` is an unambiguous method match
+        origtype = type;
+    }
+
+    // compute the type this will be cached under
+    // if we haven't selected an origtype yet, promote `type`,
+    // and then decide if it is beneficial to build a new simplesig
+    if (origtype == type) {
+        type = NULL; // don't need `type` anymore: it's equivalent to the `origtype`
+        if (makesimplesig) {
+            // reduce the complexity of rejecting this entry in the cache
+            // by replacing non-simple types with jl_any_type to build a new `type`
+            // (the only case this applies to currently due to the above logic is jl_function_type)
```

Comment:
the comment isn't new, just moved (I tried rewriting it too, but couldn't find a phrasing that was any better)